fix/feat/refactor(upgrade): make upgraded components more consistent between dynamic and static

gkalpak opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Previously, the dynamic and static versions used similar but separate implementations for upgrading components. Although there are differences, large parts of the implementations can be shared.


**What is the new behavior?**
The two versions (static and dynamic) share most of the imlementation of upgraded components (via a new helper class), which has the following effects:
- We only have to fix bugs/add features in one place and the changes propagate to both versions.
- The dynamic version is brought up-to-speed with the better implementation of the static version (which supports more features and is closer to the actual AngularJS behavior).


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ ] No
[x] Maybe (but only to the dynamic version)
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications:
With the dynamic version, upgraded components will now work more similar to how pure AngularJS components would. If an application relied on the previous behavior (e.g. not running pre-/post-linking functions, not being compiled in the correct context, not running the `$onInit()` hook, not binding required controllers to the controller instance etc), its behavior might change.


**Other information**:
This builds on top of #16627.
It also ports the fixes from #16627 to the dynamic version and fixes #11044.

gkalpak committed:
refactor(upgrade): move shareable functionality to `UpgradeHelper` class

This functionality can be potentionally re-used by the dynamic version.

gkalpak committed:
feat(upgrade): fix support for `directive.link` in upgraded components

Although, pre- and post-linking functions are correctly called during directive
linking, directives with `link.post` would throw an error. Interestingly, having
`link.pre` only or defining `link: fn` (which is an alias for `link.post: fn`)
would not throw.

This commit removes this check and allows directives with pre- and/or
post-linking functions to work.

gkalpak committed:
fix(upgrade): bring the dynamic version closer to the static one

This commit changes the dynamic version of ngUpgrade to use `UpgradeHelper`,
thus bringing its behavior (wrt upgraded components) much closer to
`upgrade/static`. Fixes/features include:

- Fix template compilation: Now takes place in the correct DOM context, instead
  of in a detached node (thus has access to required ancestors etc).
- Fix support for the `$onInit()` lifecycle hook.
- Fix single-slot transclusion (including optional transclusion and fallback
  content).
- Add support for multi-slot transclusion (inclusing optional slots and fallback
  content).
- Add support for binding required controllers to the directive's controller
  (and make the `require` behavior more consistent with AngularJS).
- Add support for pre-/post-linking functions.

(This also ports the fixes from #16627 to the dynamic version.)

Fixes #11044

mary-poppins commented:
The angular.io preview for a833b7273f35c48956f961b55f8c699e625657bc is available [here][1].

[1]: https://pr17160-a833b7273f35c48956f961b55f8c699e625657bc.ngbuilds.io/

petebacondarwin commented on packages/upgrade/src/common/upgrade_helper.ts:
```diff
@@ -169,7 +195,56 @@ export class UpgradeHelper {
+    const requiredControllers = this.resolveRequire(directiveRequire);
+
+    if (controllerInstance && this.directive.bindToController && isMap(directiveRequire)) {
+      const requiredControllersMap = requiredControllers as{[key: string]: IControllerInstance};
```
why do the `as` here rather than when the const is defined three lines above?

petebacondarwin replied:
Oh, I missed the `isMap`. If you make that function a type constraint function then you might be able to remove the type cast.

petebacondarwin replied:
Hold on, it already is a type guard. So why do you need the type cast?

gkalpak replied:
`isMap` is type-guarding `directiveRequire`. Here we are casting `requiredControllers` (because we know that its type directly depends on the type of `directiveRequire`, but TS doesn't know that).

petebacondarwin reviewed (approved):
I haven't fine tooth combed this PR, but I was happy with both commits in #16627 and the subsequent commits in this look pretty good to me.

Please squash my commit into yours @gkalpak - you did most of the work here so you deserve it.

The browserstack failure looks like it is inherited from an animations bug in master?

gkalpak committed:
fix(upgrade): fix transclusion on upgraded components

Previously, only simple, single-slot transclusion worked on upgraded components.
This commit fixes/adds support for the following:

- Multi-slot transclusion.
- Using fallback content when no transclusion content is provided.
- Destroy unused scope (when using fallback content).

Fixes #13271

gkalpak commented:
(Squashed WIP commits and rebased on master.)

mary-poppins commented:
You can preview c133f3e at https://pr17160-c133f3e.ngbuilds.io/.

mary-poppins commented:
You can preview 22a33ab at https://pr17160-22a33ab.ngbuilds.io/.

mary-poppins commented:
You can preview 6e7a73d at https://pr17160-6e7a73d.ngbuilds.io/.

jasonaden commented:
@gkalpak I created a new PR #17971 because this one won't make it through Travis as there's something wrong with upstream/pr/17160.
